### DISCUSSION TITLE
Text conditioning

### DISCUSSION THREAD

**Discussion by eonglints** [lucidrains/audiolm-pytorch#32] [Ideas]
Hey, so I'm wondering about the various options for text conditioning. At the moment, it would appear we're set up to condition using cross-attention in each of the transformers. I was wondering whether we should also consider simply pre-pending text tokens before the audio tokens? Similar to the approach taken (using image tokens) in DALL-E 1 ([paper](https://arxiv.org/abs/2102.12092), [code](https://github.com/lucidrains/DALLE-pytorch) :pray: @lucidrains), Taming Transformers ([paper](https://arxiv.org/abs/2012.09841), [code](https://github.com/CompVis/taming-transformers)) and Make-A-Scene ([paper](https://arxiv.org/abs/2203.13131), [unofficial code](https://github.com/CasualGANPapers/Make-A-Scene)) as well as Tortoise TTS ([blog post](https://nonint.com/2022/04/25/tortoise-architectural-design-doc/), [code](https://github.com/neonbjb/tortoise-tts)) with audio tokens.

I think pre-pending the text tokens would only be required for the semantic transformer as the resulting semantic tokens should encode all the phonetic content, leaving the SoundStream transformers to deal with audio decoding and further conditioning on speaker identity and other acoustic attributes.

Thoughts?

**Comment by lucidrains** [comment]
@eonglints yup! i can add the prepended way this week! (and make sure it works with @crowsonkb version of classifier free guidance)

**Reply by eonglints** [reply]
You're a legend, thank you! 🙏

**Reply by eonglints** [reply]
Just a friendly nudge :) Happy to start working on it myself too. Thanks for the other latest additions/fixes!

**Comment by eonglints** [comment]
Pretty much exactly what I've been hoping to do with text-conditioned AudioLM has just been demonstrated with a new TTS paper from Microsoft: https://valle-demo.github.io/. Pretty impressive results.
It borrows heavily from AudioLM but drops the semantic transformer and uses [Encodec](https://github.com/facebookresearch/encodec) rather than SoundStream.
<img width="958" alt="image" src="https://user-images.githubusercontent.com/48352648/211034219-9f90a69f-b65f-4792-87fc-265801f99b60.png">

**Reply by lucidrains** [reply]
yes exactly! i'll plan on make the repository compatible with the open sourced Encodec!

**Reply by lucidrains** [reply]
what can you say, attention is all we need

**Reply by eonglints** [reply]
> yes exactly! i'll plan on make the repository compatible with the open sourced Encodec!

Amazing! It's very, very similar to Soundstream - encoder-decoder with residual vector quantization.

**Reply by lucidrains** [reply]
@eonglints haha yea, to the point where i think it is just MetaAI doing an incremental paper and rebranding the soundstream paper... but that's ok, at least they released a model open source and it led to VALL-E!

**Reply by eonglints** [reply]
Absolutely! Although your work on SoundStream is still absolutely vital, as Encodec doesn't have a permissive commercial licence and doesn't have any training code. So, thanks again.

**Reply by lucidrains** [reply]
@eonglints oh i did not know that! i thought fairseq was MIT licensed though, weird that Encodec would not be the same... ok, maybe i'll continue to try to make soundstream work then (i'll look to encodec to figure out how they got around the complex valued activation not being distributed friendly)

**Reply by djqualia** [reply]
yes your work on the soundstream is much appreciated!  afaict it's working for me (making progress on a test corpus of music, and sounding better over time)... although i'm not training in a distributed setup!

**Reply by eonglints** [reply]
> @eonglints oh i did not know that! i thought fairseq was MIT licensed though, weird that Encodec would not be the same... ok, maybe i'll continue to try to make soundstream work then (i'll look to encodec to figure out how they got around the complex valued activation not being distributed friendly)

Yeah, at the moment they've opted for an Attribution-NonCommercial 4.0 International license (it's in the MetaAI repo but not in fairseq by the way). One of the authors [has said](https://github.com/facebookresearch/encodec/issues/1#issuecomment-1291941171) "We are considering the possibility of releasing this code under MIT, but it will take a bit of time to get the legal approval." So, who knows... Given how useful SoundStream-like representations are for a whole range of tasks for the audio/speech/music communities, I would say it would be great to get this version over the finishing line.

**Reply by lucidrains** [reply]
yea, same old same old. I was at the other side of the fence once

ok, I'll definitely put more work into this!

i will also avoid integrating Encodec then, given the licensing

**Reply by lucidrains** [reply]
@eonglints ok, it should work now, if one simply sets https://github.com/lucidrains/audiolm-pytorch/commit/654d49693959e08266cac624bccefeb5fe771cbc#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R194 to `True`!

**Reply by lucidrains** [reply]
although, i'm still more of a believer in cross attention, but definitely this will let us easily compare the two approaches

**Reply by eonglints** [reply]
Amazing, thank you. And yes, I think the jury's still out, especially for text-audio language models, so it will be great to compare the two approaches. I'm currently training all three Transformers and hoping to be able to share some preliminary text-conditioned results fairly soon.

**Comment by LWprogramming** [comment]
Speaking of text conditioning, it looks like Google's just published a new paper on this: https://google-research.github.io/seanet/musiclm/examples/

**Comment by djqualia** [comment]
Wow, they just demonstrated so many things I've wanted to try with this code...  I can't access the paper yet, I wonder how similar it is to what's @lucidrains has built here?

I'm excited for this MusicCaps data set too.

**Reply by lucidrains** [reply]
it is slightly different, they involve an audio CLIP model from https://arxiv.org/abs/2208.12415 I'll be building it out at https://github.com/lucidrains/musiclm-pytorch

**Comment by djqualia** [comment]
you are legendary!

yes it is very interesting how they overcame the problem of scarce training data by using a shared embedding space (Mulan).
